src/main/scala/beam/router/graphhopper/GraphHopperWrapper.scala

```diff
@@ -62,93 +64,112 @@ class GraphHopperWrapper(
+          var totalTravelTime = (responsePath.getTime / 1000).toInt
+          val ghLinkIds: IndexedSeq[Int] =
+            responsePath.getPathDetails
+              .asScala(Parameters.Details.EDGE_ID)
+              .asScala
+              .map(pd => pd.getValue.asInstanceOf[Int])
+              .toIndexedSeq
+
+          val allLinkTravelBeamTimes = responsePath.getPathDetails
+            .asScala(BeamTimeDetails.BEAM_TIME)
+            .asScala
+            .map(pd => pd.getValue.asInstanceOf[Long].toDouble / 1000.0)
+            .toIndexedSeq
+
+          val allLinkTravelBeamTimesReverse = responsePath.getPathDetails
+            .asScala(BeamTimeReverseDetails.BEAM_REVERSE_TIME)
             .asScala
-            .map(pd => pd.getValue.asInstanceOf[Int])
+            .map(pd => pd.getValue.asInstanceOf[Long].toDouble / 1000.0)
             .toIndexedSeq
 
-        val allLinkTravelTimes = responsePath.getPathDetails
-          .asScala(Parameters.Details.TIME)
-          .asScala
-          .map(pd => pd.getValue.asInstanceOf[Long].toDouble / 1000.0)
-          .toIndexedSeq
+          val allLinkTravelTimes =
+            if (Math.abs(totalTravelTime - allLinkTravelBeamTimes.sum.toInt) <= 2) {
+              allLinkTravelBeamTimes
+            } else {
+              allLinkTravelBeamTimesReverse
+            }
 
-        var linkTravelTimes: IndexedSeq[Double] = allLinkTravelTimes
-        // TODO ask why GH is producing negative travel time
-//          .map { x =>
-//            require(x > 0, "GOING BACK IN TIME")
-//            x
-//          }
-        //FIXME BECAUSE OF ADDITIONAL ZEROs WE HAVE A DISCREPANCY BETWEEN NUMBER OF LINK IDS AND TRAVEL TIMES
-          .take(ghLinkIds.size)
+          val linkTravelTimes: IndexedSeq[Double] = allLinkTravelTimes
+          // TODO ask why GH is producing negative travel time
+          //          .map { x =>
+          //            require(x > 0, "GOING BACK IN TIME")
+          //            x
+          //          }
+          //FIXME BECAUSE OF ADDITIONAL ZEROs WE HAVE A DISCREPANCY BETWEEN NUMBER OF LINK IDS AND TRAVEL TIMES
+            .take(ghLinkIds.size)
 
-        if (allLinkTravelTimes.size > ghLinkIds.size) {
-          allLinkTravelTimes.drop(ghLinkIds.size).foreach { time =>
-            totalTravelTime -= time.toInt
+          if (allLinkTravelTimes.size > ghLinkIds.size) {
+            allLinkTravelTimes.drop(ghLinkIds.size).foreach { time =>
+              totalTravelTime -= time.toInt
+            }
           }
-        }
 
-        if (ghLinkIds.size < 2) {
-          // An empty path by GH's definition. But we still want it to be from a link to a link.
-          val snappedPoint = graphHopper.getLocationIndex.findClosest(origin.getY, origin.getX, EdgeFilter.ALL_EDGES)
-          val edgeId = snappedPoint.getClosestEdge.getEdge * 2
+          if (ghLinkIds.size < 2) {
+            None
+          } else {
+            linkIds = ghLinkIds
```

Comment:
nitpick: Is it some sort of conversion from GH link ids to R5 link ids? Should it be moved to separate method

Reply:
fixed